accepted/2025/indicate-analyzer-assets.md

```diff
@@ -0,0 +1,93 @@
+
+- Author Name: [kfikadu](https://github.com/kfikadu)
+- GitHub Issue: [Issue #6279](https://github.com/NuGet/Home/issues/6279)
+
+## Summary
+
+`project.assets.json` does not currently indicate which analyzers are active, and analyzers are added even when `PrivateAssets` or `ExcludeAssets` should prevent this.
+This leads to analyzers being included in projects, which are not the expected behavior, especially when defaults for private/excluded assets are not respected
+
+## Motivation
+
+Currently, analyzer assets in NuGet packages do not respect asset filtering options like ExcludeAssets and PrivateAssets, unlike other asset types (compile, runtime).
+Analyzers are included based on pattern matching in package files, not by explicit references in project.assets.json.
+Developers who try to exclude analyzers using asset filters still see analyzers loaded and executed during builds, leading to unexpected warnings or errors.
+This causes wasted time, frustration, and confusion, as configuration changes appear ineffective and troubleshooting becomes more difficult.
+Package authors cannot reliably control analyzer distribution or exclusion in downstream projects, making support and package design more complex.
+Real-world feedback shows users resorting to custom MSBuild scripts, manual artifact modification, or abandoning packages due to inability to control analyzer inclusion.
+
+## Explanation
+
+### Functional explanation
+
+Analyzer assets in NuGet packages will now respect asset filtering options like ExcludeAssets and PrivateAssets, just as other asset types do.
+- When excluding analyzers from a package, they will no longer be included or executed during the project's build.
+- Configurations will accurately control which analyzers are present, eliminating unexpected analyzer warnings or errors.
+- Both package authors and consumers will have clear, reliable control over analyzer inclusion without needing custom workarounds.
+
+This feature will initially be behind a feature flag to avoid doing a breaking change.
+
+### Technical explanation
+
+#### NuGet changes
+
+Update the NuGet restore process to:
+- Identify analyzer DLLs in each package.
+- Filter out analyzers that are excluded by PrivateAssets/ExcludeAssets
+- Add an "analyzers" group under the appropriate "targets" node for each package in project.assets.json.
+
+Example Output
+```json
+"targets": {
+  ".NETCoreApp,Version=v8.0": {
+    "My.Analyzer.Package/1.0.0": {
+      "analyzers": [
+        "analyzers/dotnet/cs/MyAnalyzer.dll"
+      ]
+    }
+  }
+}
+```
+
+#### SDK Changes
+
+Update SDK asset resolution logic (e.g., ResolvePackageAssets.cs) to:
+- Read analyzer assets directly from the "analyzers" group in "targets", instead of scanning all files in "libraries".
+- Only load analyzers are listed in this group.
+- If the "analyzers" group is absent (e.g., from older lock files), fall back to legacy scanning to preserve compatibility
```

Comment:
We should probably "teach" the SDK how to skip the scanning when the feature is enabled but it just let to no analyzers being selected since analyzers don't exist in any of the packages.